main.py

```diff
@@ -833,85 +836,272 @@ def on_link_down(self, event):
-                            "Ignore Failover path for "
-                            f"{evc} due to error: {err}"
-                        )
-                        evcs_normal.append(evc)
+                    )):
+                        undeploy.append(evc)
+                    elif all((
+                        not evc.is_affected_by_link(link),
+                        evc.failover_path,
+                        evc.is_failover_path_affected_by_link(link)
+                    )):
+                        clear_failover.append(evc)
+                    else:
                         continue
-                    for dpid, flows in dpid_flows.items():
-                        switch_flows.setdefault(dpid, [])
-                        switch_flows[dpid].extend(flows)
-                    evcs_with_failover.append(evc)
-                    failover_event_contents[evc.id] = map_evc_event_content(
-                        evc,
-                        flows={k: v.copy() for k, v in switch_flows.items()}
-                    )
-                elif evc.is_failover_path_affected_by_link(link):
-                    evc.old_path = evc.failover_path
-                    evc.failover_path = Path([])
-                    check_failover.append(evc)
-
-        if failover_event_contents:
-            emit_event(self.controller, "failover_link_down",
-                       content=deepcopy(failover_event_contents))
-        send_flow_mods_event(self.controller, switch_flows, 'install')
-
-        for evc in evcs_normal:
-            emit_event(
-                self.controller,
-                "evc_affected_by_link_down",
-                content={"link": link} | map_evc_event_content(evc)
-            )
 
-        evcs_to_update = []
-        for evc in evcs_with_failover:
-            evcs_to_update.append(evc.as_dict())
-            log.info(
-                f"{evc} redeployed with failover due to link down {link.id}"
-            )
-        for evc in check_failover:
-            evcs_to_update.append(evc.as_dict())
+                    exit_stack.push(sub_stack.pop_all())
 
-        self.mongo_controller.update_evcs(evcs_to_update)
+            # Swap from current path to failover path
 
-        emit_event(
-            self.controller,
-            "cleanup_evcs_old_path",
-            content={"evcs": evcs_with_failover + check_failover}
-        )
+            success, failure = self.execute_swap_to_failover(swap_to_failover)
+
+            clear_failover.extend(success)
+
+            evcs_to_update.update((evc.id, evc) for evc in success)
+
+            undeploy.extend(failure)
+
+            # Clear out failover path
+
+            success, failure = self.execute_clear_failover(clear_failover)
+
+            evcs_to_update.update((evc.id, evc) for evc in success)
+
+            undeploy.extend(failure)
+
+            # Undeploy the evc, schedule a redeploy
+
+            success, failure = self.execute_undeploy(undeploy)
```

Comment:
Should we log the `failure` list as error after this statement if it's not empty? Since after this statement we're not doing anything with the `failure` list. The `prepare_*` errors will be extremely rare, but if they do happen, would be cool to also have a log error for us to debug and understand in the future if it ever happens.

Reply:
+1 for having more log statements in case of failure (of course, we have to measure the compromise with performance impact, but one log message at least is beneficial).

Reply:
Logging has been added in. Also E2E tests are now passing.